target/linux/lantiq/patches-6.6/0151-lantiq-ifxmips_pcie-use-of.patch

```diff
@@ -260,15 +260,14 @@ Signed-off-by: Eddi De Pieri <eddi@depieri.net>
 +    if (IS_ERR(ltq_rcu_regmap))
 +        return PTR_ERR(ltq_rcu_regmap);
 +
-+    pcie_reset_gpio = of_get_named_gpio(node, "gpio-reset", 0);
-+    if (gpio_is_valid(pcie_reset_gpio)) {
-+        int ret = devm_gpio_request(&pdev->dev, pcie_reset_gpio, "pcie-reset");
-+        if (ret) {
-+            dev_err(&pdev->dev, "failed to request gpio %d\n", pcie_reset_gpio);
-+            return ret;
-+        }
-+        gpio_direction_output(pcie_reset_gpio, 1);
++    reset_gpio = devm_gpiod_get_optional(&pdev->dev, "reset",
++                                         GPIOD_OUT_LOW);
++    error = PTR_ERR_OR_ZERO(reset_gpio);
```

Comment:
PTR_ERR_OR_ZERO implies it's not optional at all.

Reply:
```
 * Return: The error code within @ptr if it is an error pointer; 0 otherwise.
 */
static inline int __must_check PTR_ERR_OR_ZERO(__force const void *ptr)
{
	if (IS_ERR(ptr))
		return PTR_ERR(ptr);
	else
		return 0;
}
```
`devm_gpiod_get_optional()` will return 0 when no GPIO assigned. Then `PTR_ERR_OR_ZERO(0)` will return 0, too. I think the logic is correct here.

I guess you confused `PTR_ERR_OR_ZERO()` with `IS_ERR_OR_NULL()`.
```
 * Like IS_ERR(), but also returns true for a null pointer.
 */
static inline bool __must_check IS_ERR_OR_NULL(__force const void *ptr)
{
	return unlikely(!ptr) || IS_ERR_VALUE((unsigned long)ptr);
}
```